clang/lib/Sema/SemaTemplateDeductionGuide.cpp

```diff
@@ -936,19 +940,21 @@ Expr *buildIsDeducibleConstraint(Sema &SemaRef,
         Context.DeclarationNames.getCXXDeductionGuideName(AliasTemplate));
   };
 
+  TemplateDecl *TD = DeducingTemplate ? DeducingTemplate : AliasTemplate;
+
   SmallVector<TypeSourceInfo *> IsDeducibleTypeTraitArgs = {
       Context.getTrivialTypeSourceInfo(
           Context.getDeducedTemplateSpecializationType(
-              TemplateName(AliasTemplate), /*DeducedType=*/QualType(),
+              TemplateName(TD), /*DeducedType=*/QualType(),
```

Comment:
```cpp
template <typename T> struct B {
  B(T);
};
template <typename T> struct C : public B<T> {
  using B<T>::B;
};

C c(1);

// implicit code
// template <typename> class CC;
//
// template <typename T>
// class CC<B<T>> {
// public:
//    typedef C<T> type;
// };
```
If I read the code correctly,  we will build the `__is_deducible(C, typename CC<B<T>>::type)` for the deduction guide of `C` (`auto (T) -> typename CC<B<T>>::type`).  I'm not sure this is correct.

Intuitively, this constraint is not necessary, if `CC<B<T>>::type` is valid, it must be the type of `C<T>`, then `is_deducible(C, C<T>)` is always true.

Unfortunately, the standard doesn't specify the transformation of the associated constraint (oversight?). If we only replace the return type, then the constraint will be `is_deducible(B,  typename CC<B<T>>::type)`  which is always false. 

Perhaps we should always drop the `is_deducible` constraint for this case.

Reply:
I think you're right that the `is_deducible` constraint is technically unnecessary-- as far as I can tell, any case where this constraint would fail would cause the `CC<R>::type` instantiation to fail before the constraint is evaluated.
There are some other caveats with removing it, however. Consider this example:

```c++
template<typename T>
struct Base1 { };

template<typename T>
struct Base2 { };

template<typename T = int>
struct Derived : public Base1<T>, Base2<T> {
    using Base1<T>::Base1;
    using Base2<T>::Base2;
};

Derived d;
```
https://godbolt.org/z/4YcdvTMov

Under [[over.match.best.general]p2.5](https://eel.is/c++draft/over.match.best.general#2.5) (and more specifically [[temp.func.order]p6.4](https://eel.is/c++draft/temp.func.order#6.4)), the bases' deduction guides are more specialized than any from `Derived` (due to the `is_deducible` constraint), and so the implementation as it is sees this as ambiguous.

Since the standard doesn't specify what to do with the deducible constraint, and unless there's an oversight here, I assume this is what it wants to happen.

Given that [[over.match.best.general]p2.10](https://eel.is/c++draft/over.match.best.general#2.10) would otherwise give priority to the `Derived` deduction guide, it's possible there's an oversight, but I'm not sure if that warrants removing the deducible constraint yet.

Reply:
Per https://eel.is/c++draft/over.match.class.deduct#1.sentence-7, 
I think we imply we should copy the constraints of the made up alias. WDYT?

Feel free to open an issue here though https://github.com/cplusplus/CWG/issues

Reply:
Based on the current writing of the standard I think we should be copying all of the constraints over, including the deducible constraint. However, it's probably not the intent to break existing code as in the above example.

I have filed https://github.com/cplusplus/CWG/issues/607 . Is the best course of action in the meantime to update the implementation in this PR to accept that above code, or leave it as breaking until CWG responds?

Reply:
I believe I forgot to acknowledge this comment. I am skipping the construction of the deducible constraint as of [0245459](https://github.com/llvm/llvm-project/pull/98788/commits/024545921b5ac4c8e9fa4781b492cf719877cb05)